### PR TITLE
CLDR-14557 Modernize look up code or xpath (browse); what-is

### PR DESCRIPTION
**Pull request by btangmu**
-Add Whatis.java and WhatisResponse.java, not done yet

-Run npx prettier on all .vue files

##### Checklist

- [x] Issue filed: https://unicode-org.atlassian.net/browse/CLDR-14557
- [x] Updated PR title and link in previous line to include Issue number

### CODE REVIEW

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/Whatis.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+        }
+        WhatisResponse what = new WhatisResponse(q);
+        if (q.startsWith("\\u") || q.startsWith("=")) {
+            searchPathsAndValues(what); // doesn't use loc
+        } else {
+            searchStandardCodesExact(what, loc);
+            searchStandardCodesFull(what, loc);
+        }
+        return Response.ok().entity(what).build();
+    }
```
Comment (btangmu) [inline tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/Whatis.java:64]: @srl295 I've imitated LoginResponse.java for `entity(what)` here and `@Schema(implementation = WhatisResponse.class)` above

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/WhatisResponse.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import org.unicode.cldr.util.CLDRLocale;
+
+public final class WhatisResponse {
+    @Schema(description = "query string")
+    public String q;
+    public String q2 = null;
+    public String err = null;
+    public List<XpathResult> xpath = null;
+    public List<StandardCodeResult> exact = null;
+    public List<StandardCodeResult> full = null;
```
Comment (btangmu) [inline tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/WhatisResponse.java:16]: I'm getting this when exact is non-null and non-empty:
```
[ERROR   ] Unable to serialize property 'exact' from org.unicode.cldr.web.api.WhatisResponse
[ERROR   ] Generating incomplete JSON
[ERROR   ] Problem with writing the data, class org.unicode.cldr.web.api.WhatisResponse, ContentType: application/json
[err] java.lang.Throwable: ANDY
```

File: tools/cldr-apps/js/src/views/AboutPanel.vue [hunk 3]
```diff
@@ -1,73 +1,70 @@
 <template>
-    <div id="home">
-        <p class='hang'>
```
Comment (btangmu) [inline tools/cldr-apps/js/src/views/AboutPanel.vue:3]: When I make a real (not draft) pr, I can back off on the changes that are only .vue formatting by npx prettier -- let's coordinate on that when convenient

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/WhatisResponse.java (hunk 2 above)
Comment (srl295) [reply tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/WhatisResponse.java:16]: yes ANDY commented here https://github.com/OpenLiberty/open-liberty/issues/15845#issuecomment-791722057

What you need to do is to add a `public StandardCodeResult[] getExact() {}` function,  using `toArray` - see for example: 

https://github.com/unicode-org/cldr/pull/1090/files#diff-1d25604a4ccc92b62c7dfc8f79474238ce51a65ac3812e38ce663bacfb4a5b4aR854-R858

a public `getExact()` function causes `exact` to show up as a property.

File: tools/cldr-apps/js/src/views/AboutPanel.vue (hunk 3 above)
Comment (srl295) [reply tools/cldr-apps/js/src/views/AboutPanel.vue:3]: or just make a separate PR that's just the formatting?

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/WhatisResponse.java (hunk 2 above)
Comment (btangmu) [reply tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/WhatisResponse.java:16]: I don't actually see what you're referring to in that PR; anyway I'll try adding `getExact()`, thanks!

It's mysterious why or when a getter is needed -- converting a Java List to a json array would seem like a natural default. If it takes so much code to convert an object to json, then the advantages over using JSONObject aren't so clear... Do you happen to know a good guide to using javax.ws.rs.core.Response? Other than https://docs.oracle.com/javaee/7/api/javax/ws/rs/core/Response.html ?

Comment (srl295) [reply tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/WhatisResponse.java:16]: I don't know off hand no

Comment (srl295) [reply tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/WhatisResponse.java:16]:
```java
       List<ReviewEntry> entries;

        // for serialization
         public ReviewEntry[] getEntries() {
             return entries.toArray(new ReviewEntry[entries.size()]);
         }
```

Comment (srl295) [reply tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/WhatisResponse.java:16]: I think the situation is that `List` isn't a  primitive JSON type ,and won't round-trip from an array.

Comment (btangmu) [reply tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/WhatisResponse.java:16]: Here's my latest WhatisResponse.java:
```
package org.unicode.cldr.web.api;

import java.util.ArrayList;
import java.util.List;

import org.eclipse.microprofile.openapi.annotations.media.Schema;
import org.unicode.cldr.util.CLDRLocale;

public final class WhatisResponse {
    @Schema(description = "query string")
    public String q;
    public String q2 = null;
    public String err = null;
    public List<XpathResult> xpath = null;
    public List<StandardCodeResult> exact = null;
    public List<StandardCodeResult> full = null;
    public int pathsSearched = 0;
    public int localesSearched = 0;

    public WhatisResponse(String queryString) {
        q = queryString;
    }

    public XpathResult[] getXpath() {
        System.out.println("Hello my name is getXpath");
        return xpath.toArray(new XpathResult[xpath.size()]);
    }

    public StandardCodeResult[] getExact() {
        System.out.println("Hello my name is getExact");
        return exact.toArray(new StandardCodeResult[exact.size()]);
    }

    public StandardCodeResult[] getFull() {
        System.out.println("Hello my name is getFull");
        return full.toArray(new StandardCodeResult[full.size()]);
    }

    class XpathResult {
        public String loc;
        public String name;
        public List<PathAndValue> matches = new ArrayList<>();

        public XpathResult(CLDRLocale locale) {
            loc = locale.toString();
            name = locale.getDisplayName();
        }

        public PathAndValue[] getMatches() {
            System.out.println("Hello my name is getMatches");
            return matches.toArray(new PathAndValue[matches.size()]);
        }
    }

    class PathAndValue {
        public String path;
        public String value;

        public PathAndValue(String p, String v) {
            path = p;
            value = v;
        }
    }

    class StandardCodeResult {
        public String type;
        public String code;
        public List<StandardCodeResultDataItem> data = new ArrayList<>();

        public StandardCodeResult(String t, String c) {
            type = t;
            code = c;
        }

        public StandardCodeResultDataItem[] getData() {
            System.out.println("Hello my name is getData");
            return data.toArray(new StandardCodeResultDataItem[data.size()]);
        }
    }

    class StandardCodeResultDataItem {
        public String s;
        public boolean isWinner;

        public StandardCodeResultDataItem(String str, boolean win) {
            s = str;
            isWinner = win;
        }
    }
}
```

And here's the console:

```
Hello my name is getExact
[ERROR   ] Unable to serialize property 'exact' from org.unicode.cldr.web.api.WhatisResponse
[ERROR   ] Generating incomplete JSON
[ERROR   ] Problem with writing the data, class org.unicode.cldr.web.api.WhatisResponse, ContentType: application/json
[err] java.lang.Throwable: ANDY
[err] 	at org.apache.cxf.jaxrs.utils.JAXRSUtils.logMessageHandlerProblem(JAXRSUtils.java:2055)
[err] 	at [internal classes]
[err] 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:686)
[err] 	at com.ibm.websphere.jaxrs.server.IBMRestServlet.service(IBMRestServlet.java:96)
[err] 	at [internal classes]
[err] 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
[err] 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
[err] 	at java.base/java.lang.Thread.run(Thread.java:853)
[WARNING ] Interceptor for {http://api.web.cldr.unicode.org/}About has thrown exception, unwinding now
Unable to serialize property 'exact' from org.unicode.cldr.web.api.WhatisResponse
[WARNING ] Could not create a secure Stax XMLInputFactory.  Found class com.sun.xml.internal.stream.XMLInputFactoryImpl.  Suggest Woodstox 4.2.0 or newer.
```

I have no idea whether `Woodstox` is related...

Comment (srl295) [reply tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/WhatisResponse.java:16]: don't make these public. 
```java
    public List<StandardCodeResult> exact = null;
    public List<StandardCodeResult> full = null;
```

Comment (btangmu) [reply tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/WhatisResponse.java:16]: > don't make these public.

I changed them to private. The `getExact` function does get called, and then the same error still occurs: `Unable to serialize property 'exact'.`

**Comment by srl295** [comment]
Is there some complexity in the standard code class

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/WhatisResponse.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        public PathAndValue(String p, String v) {
+            path = p;
+            value = v;
+        }
+    }
+
+    class StandardCodeResult {
+        public String type;
+        public String code;
+        public List<StandardCodeResultDataItem> data = new ArrayList<>();
```
Comment (srl295) [inline tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/WhatisResponse.java:48]: @btangmu this also needs to be an object array

Comment (btangmu) [reply tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/WhatisResponse.java:48]: > this also needs to be an object array

Yes, I made all the List members private with getters; still the same error. Here's the latest:
```
package org.unicode.cldr.web.api;

import java.util.ArrayList;
import java.util.List;

import org.eclipse.microprofile.openapi.annotations.media.Schema;
import org.unicode.cldr.util.CLDRLocale;

public final class WhatisResponse {
    @Schema(description = "query string")
    public String q;
    public String q2 = null;
    public String err = null;
    public int pathsSearched = 0;
    public int localesSearched = 0;

    private List<XpathResult> xpath = null;
    private List<StandardCodeResult> exact = null;
    private List<StandardCodeResult> full = null;

    public WhatisResponse(String queryString) {
        q = queryString;
    }

    public XpathResult[] getXpath() {
        System.out.println("Hello my name is getXpath");
        return xpath.toArray(new XpathResult[xpath.size()]);
    }

    public StandardCodeResult[] getExact() {
        System.out.println("Hello my name is getExact");
        return exact.toArray(new StandardCodeResult[exact.size()]);
    }

    public StandardCodeResult[] getFull() {
        System.out.println("Hello my name is getFull");
        return full.toArray(new StandardCodeResult[full.size()]);
    }

    class XpathResult {
        public String loc;
        public String name;
        private List<PathAndValue> matches = new ArrayList<>();

        public XpathResult(CLDRLocale locale) {
            loc = locale.toString();
            name = locale.getDisplayName();
        }

        public PathAndValue[] getMatches() {
            System.out.println("Hello my name is getMatches");
            return matches.toArray(new PathAndValue[matches.size()]);
        }

        public void addMatch(PathAndValue pathAndValue) {
            matches.add(pathAndValue);
        }
    }

    class PathAndValue {
        public String path;
        public String value;

        public PathAndValue(String p, String v) {
            path = p;
            value = v;
        }
    }

    class StandardCodeResult {
        public String type;
        public String code;
        private List<StandardCodeResultDataItem> data = new ArrayList<>();

        public StandardCodeResult(String t, String c) {
            type = t;
            code = c;
        }

        public StandardCodeResultDataItem[] getData() {
            System.out.println("Hello my name is getData");
            return data.toArray(new StandardCodeResultDataItem[data.size()]);
        }

        public void addData(StandardCodeResultDataItem item) {
            data.add(item);
        }
    }

    class StandardCodeResultDataItem {
        public String s;
        public boolean isWinner;

        public StandardCodeResultDataItem(String str, boolean win) {
            s = str;
            isWinner = win;
        }
    }

    public void addExact(StandardCodeResult r) {
        if (exact == null) {
            exact = new ArrayList<>();
        }
        exact.add(r);
    }

    public void addFull(StandardCodeResult r) {
        if (full == null) {
            full = new ArrayList<>();
        }
        full.add(r);
    }

    public void addXpath(XpathResult r) {
        if (xpath == null) {
            xpath = new ArrayList<>();
        }
        xpath.add(r);
    }
}
```

I wonder if there's a debug mode for the serialization library, or a way to step through its code, to get more clues about where things go wrong. I'd also love to find some documentation clarifying the requirements for serialization, such as when does public/private actually matter, when are `@Schema` or other annotations or metadata required.

This article might have useful ideas: https://www.baeldung.com/jackson-custom-serialization

It's not even clear to me what the relevant technology is here: Jackson? Swagger?

Thanks!

**Comment by srl295** [comment]
Swagger is just documentation. May be Jackson.

**Comment by srl295** [comment]
@yumaoka any ideas on debugging serialization?

**Comment by btangmu** [comment]
Seemingly the error occurs within `javax.ws.rs.core.Response.ResponseBuilder.entity(Object)` so a key role is played by JAX-RS. Docs:

* https://www.ibm.com/support/knowledgecenter/SSEQTP_8.5.5/com.ibm.websphere.base.iseries.doc/ae/cwbs_jaxrs_overview.html

* https://www.ibm.com/support/knowledgecenter/SSEQTP_8.5.5/com.ibm.websphere.base.iseries.doc/ae/twbs_jaxrs_jsoncontent.html

**Comment by btangmu** [comment]
"By default, JSONB ignores properties with a non public access. All public properties - either public fields or non public fields with public getters are serialized into JSON text."
http://json-b.net/docs/user-guide.html#ignoring-properties

**Comment by btangmu** [comment]
See https://github.com/unicode-org/cldr/pull/1113

**Pull request closed without merging**